devices/VectorsCollectionWrapper/src/VectorsCollectionWrapper.cpp

```diff
@@ -0,0 +1,177 @@
+        return false;
+    }
+
+    if (remotePortNames.size() != remoteVarNames.size())
+    {
+        log()->error("{} Lists \"remote_port_names\" and \"remote_var_names\""
+                     "must follow same size and order.", logPrefix);
+        return false;
+    }
+
+    bool ok = setupPortManager(remotePortNames, remoteVarNames, m_portPrefix, m_portManager);
+    if (!ok)
+    {
+        log()->error("{} Unable to configure port manager.", logPrefix);
+        return false;
+    }
+
+    if (!params->getParameter("output_port_name", m_wrapperPortName))
+    {
+        log()->error("{} Unable to get required parameter \"output_port_name\".", logPrefix);
+        return false;
+    }
+
+    if (m_wrapperPortName[0] != '/')
+    {
+        log()->error("{} \"output_port_name\" should begin with character \'/\'.", logPrefix);
+        return false;
+    }
+
+    m_wrapperPort.open(m_wrapperPortName);
+
+    if (ok)
+    {
+        return start();
+    }
+
+    return ok;
+}
+
+bool VectorsCollectionWrapper::setupPortManager(const std::vector<PortName>& portNames,
+                                                const std::vector<VarOutName>& varNames,
+                                                const std::string& portPrefix,
+                                                std::unordered_map<PortName, VectorPortData>& portManager)
+{
+    constexpr auto logPrefix = "[VectorsCollectionWrapper::setupPortManager]";
+    std::size_t idx{0};
+    for (const auto& remote : portNames)
+    {
+        const auto& varOutName = varNames[idx];
+        std::string inPort{portPrefix + varOutName + ":i"};
+        if (remote[0] != '/')
+        {
+            log()->error("{} Every port in \"remote_port_names\" "
+                         "should begin with character \'/\'.", logPrefix);
+            return false;
+        }
+
+        // open input port
+        portManager[remote].inputPortName = inPort;
+        portManager[remote].varName = varOutName;
+        portManager[remote].port.open(inPort);
+        if (!yarp::os::Network::connect(remote, inPort))
+        {
+            log()->error("{} Unable to connect the ports {} and {}",
+                         logPrefix, remote, inPort);
+            return false;
+        }
+
+        idx++;
+    }
+    return true;
+}
+
+void VectorsCollectionWrapper::run()
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    m_placeHolderVec.clear();
+    auto& data = m_wrapperPort.prepare();
+    data.vectors.clear();
+    // get the data from the ports
+    // wrap in VectorsCollection and send
+    for (auto& [remote, portData] : m_portManager)
+    {
+        yarp::sig::Vector* vec{nullptr};
+        vec = portData.port.read(/*shouldWait = */ false);
+        if (vec)
+        {
+            m_placeHolderVec = *vec;
+            data.vectors[portData.varName].assign(m_placeHolderVec.data(),
+                                                  m_placeHolderVec.data() + m_placeHolderVec.size());
```

Comment:
Why not using directly ``vec`` to copy the data into ``data.vectors[portData.varName]``? Right now there are two consequent copies.

Reply:
Fixed in https://github.com/ami-iit/bipedal-locomotion-framework/pull/512/commits/89f521664c8519b88327d7655706cff550c4715a.